includes/class-reader-activation.php

```diff
@@ -0,0 +1,328 @@
+	 *
+	 * @return \WP_User|\WP_Error The authenticated reader or WP_Error if authentication failed.
+	 */
+	private static function set_current_reader( $user_id, $verify_email = false ) {
+		$user_id = \absint( $user_id );
+		if ( empty( $user_id ) ) {
+			return new \WP_Error( 'newspack_authenticate_invalid_user_id', __( 'Invalid user id.', 'newspack' ) );
+		}
+
+		$user = \get_user_by( 'id', $user_id );
+		if ( ! $user || \is_wp_error( $user ) || ! self::is_user_reader( $user ) ) {
+			return new \WP_Error( 'newspack_authenticate_invalid_user', __( 'Invalid user.', 'newspack' ) );
+		}
+
+		if ( true === $verify_email ) {
+			self::verify_reader_email( $user );
+		}
+
+		\wp_clear_auth_cookie();
+		\wp_set_current_user( $user->ID );
+		\wp_set_auth_cookie( $user->ID );
+		\do_action( 'wp_login', $user->user_login, $user );
+
+		return $user;
+	}
+
+	/**
+	 * Register a reader given its email.
+	 *
+	 * Due to authentication or auth intention, this method should be used
+	 * preferably on POST or API requests to avoid issues with caching.
+	 *
+	 * @param string $email        Email address.
+	 * @param string $display_name Reader display name to be used on account creation.
+	 * @param bool   $authenticate Whether to authenticate after registering. Default to true.
+	 * @param bool   $notify       Whether to send email notification to the reader. Default to true.
+	 *
+	 * @return int|false|\WP_Error The created user ID in case of registration, false if the user already exists, or a WP_Error object.
+	 */
+	public static function register_reader( $email, $display_name = '', $authenticate = true, $notify = true ) {
+		if ( ! self::is_enabled() ) {
+			return new \WP_Error( 'newspack_register_reader_disabled', __( 'Registration is disabled.', 'newspack' ) );
+		}
+
+		if ( \is_user_logged_in() ) {
+			return new \WP_Error( 'newspack_register_reader_logged_in', __( 'Cannot register while logged in.', 'newspack' ) );
+		}
+
+		$email = \sanitize_email( $email );
+
+		if ( empty( $email ) ) {
+			return new \WP_Error( 'newspack_register_reader_empty_email', __( 'Please enter a valid email address.', 'newspack' ) );
+		}
+
+		self::set_auth_intention( $email );
+
+		$existing_user = \get_user_by( 'email', $email );
+		if ( \is_wp_error( $existing_user ) ) {
+			return $existing_user;
+		}
+
+		$user_id = false;
+
+		if ( ! $existing_user ) {
+			/**
+			 * Create new reader.
+			 */
+			if ( empty( $display_name ) ) {
+				$display_name = explode( '@', $email, 2 )[0];
+			}
+
+			$random_password = \wp_generate_password();
+
+			if ( function_exists( '\wc_create_new_customer' ) ) {
+				/**
+				 * Create WooCommerce Customer if possible.
+				 *
+				 * Email notification for WooCommerce is handled by the plugin.
+				 */
+				$user_id = \wc_create_new_customer( $email, $email, $random_password, [ 'display_name' => $display_name ] );
+			} else {
+				$user_id = \wp_insert_user(
+					[
+						'user_login'   => $email,
```

Comment:
Suggestion: base user login on the display name, if provided.

Reply:
This would create unnecessary conflict with potential duplicate logins.